### PR TITLE
vm: explicit error checks in ecrecover #505

### LINKED ISSUE #505: SEC-28 Add validations to Go EC recover utility function
https://github.com/ethereum/go-ethereum/blob/develop/vm/address.go#L64

Remove the defer and handle all possible errors within the function.
Verify expected length of signature before slicing.
Verify value of the v parameter before appending it.

### PR DESCRIPTION
**Pull request by ebuchman**
What's the proper way to run the tests for something like this?

### CODE REVIEW

File: vm/address.go [hunk 1]
```diff
@@ -62,14 +62,30 @@ func ripemd160Func(in []byte) []byte {
 }
 
 func ecrecoverFunc(in []byte) []byte {
-	// In case of an invalid sig. Defaults to return nil
-	defer func() { recover() }()
-
+	// "in" is (hash, v, r, s), each 32 bytes
+	// but for ecrecover we want (r, s, v)
+	l := 32 * 4
```
Comment (obscuren) [inline vm/address.go]: No need

File: vm/address.go [hunk 2]
```diff
@@ -62,14 +62,30 @@ func ripemd160Func(in []byte) []byte {
...
+	if len(in) < l {
+		return nil
+	}
 	hash := in[:32]
-	v := common.BigD(in[32:64]).Bytes()[0] - 27
-	sig := append(in[64:], v)
+	encodedV := in[32:64]
+	rS := in[64:]
+	// v is only a bit, but comes as 32 bytes from vm, so cast to big int and grab the 1st byte
+	v := common.BigD(encodedV).Bytes()[0] - 27
```
Comment (obscuren) [inline vm/address.go]: I know this was there before, but just take that byte w/o casting to `big.Int`

File: vm/address.go [hunk 3]
```diff
@@ -62,14 +62,30 @@ func ripemd160Func(in []byte) []byte {
...
+	if !(v == 0 || v == 1) {
+		return nil
+	}
+	sig := append(rS, v)
```
Comment (obscuren) [inline vm/address.go]: Pass `sig` directly to `append`

File: vm/address.go [hunk 4]
```diff
@@ -62,14 +62,30 @@ func ripemd160Func(in []byte) []byte {
...
+	// secp256.go returns either nil or 65 bytes
+	if pubKey == nil || len(pubKey) != 65 {
+		return nil
+	}
+	// the first byte is bitcoin heritage
+	pubKey = pubKey[1:]
 
-	return common.LeftPadBytes(crypto.Sha3(crypto.Ecrecover(append(hash, sig...))[1:])[12:], 32)
+	return common.LeftPadBytes(crypto.Sha3(pubKey)[12:], 32)
 }
```
Comment (obscuren) [inline vm/address.go]:
```go
return common.LeftPadBytes(crypto.Sha3(pubkey[1:])[12:], 32)
```

**Comment by obscuren** [comment]
In fact ignore those changes. I'll put them directly in to the other branch (`conversion`) which takes fixed size length array instead of slices.

**Comment by ebuchman** [comment]
So ecrecover will explicitly take a length 32*4 array? Do you want me to make those changes and PR against `conversion`? I'm also not sure still how to run the vm tests - would be great if there was a quick readme on that somewhere.

**Comment by obscuren** [comment]
you run the tests with `go test` in `tests/vm` :-) you can rebase on develop; `conversion` has been merged.

**Comment by ebuchman** [comment]
Cool that's easy enough. Vm tests pass. Rebased and made changes re some of your comments. But why do you say the first length check is not needed? and why is the signature backwards in the first place (ie. why is V first)? And these native contracts still take a byte slice

**Comment by obscuren** [comment]
You mean my comment on `l := 32 * 4`? Just `if len(in) < 128 {`

Or define a const

**Comment by ebuchman** [comment]
Right. Thought the `32*4` might be more explicit. Changed.

**Comment by obscuren** [comment]
Sorry I haven't merged this yet. I'll give this some attention asap.

**Comment by ebuchman** [comment]
Please, no need to apologize - you a busy guy. But thanks!

**Comment by obscuren** [comment]
Removed the "closes" bit because we don't want the https://github.com/ethereum/go-ethereum/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+SEC to be closed automatically

**Pull request merged**